src/common/meta/src/key.rs

```diff
@@ -221,12 +511,282 @@ impl_table_meta_value! {
+            .unwrap();
+        // if metadata was already created, it should be ok.
+        table_metadata_manager
+            .create_table_metadata(table_info.clone(), region_routes.clone())
+            .await
+            .unwrap();
+        let mut modified_region_routes = region_routes.clone();
+        modified_region_routes.push(region_route);
+        // if remote metadata was exists, it should return an error.
+        assert!(table_metadata_manager
+            .create_table_metadata(table_info, modified_region_routes)
+            .await
+            .is_err());
+    }
+
+    #[tokio::test]
+    async fn test_delete_table_metadata() {
+        let mem_kv = Arc::new(MemoryKvBackend::default());
+        let table_metadata_manager = TableMetadataManager::new(mem_kv);
+        let table_info: RawTableInfo = new_test_table_info().into();
+        let region_route = new_test_region_route();
+        let region_routes = vec![region_route.clone()];
+        // creates metadata.
+        table_metadata_manager
+            .create_table_metadata(table_info.clone(), region_routes.clone())
+            .await
+            .unwrap();
+        let table_info_value = TableInfoValue::new(table_info);
+        // deletes metadata.
+        table_metadata_manager
+            .delete_table_metadata(table_info_value.clone(), region_routes.clone())
+            .await
+            .unwrap();
+        // if metadata was already deleted, it should be ok.
+        table_metadata_manager
+            .delete_table_metadata(table_info_value.clone(), region_routes.clone())
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_rename_table() {
+        let mem_kv = Arc::new(MemoryKvBackend::default());
+        let table_metadata_manager = TableMetadataManager::new(mem_kv);
+        let table_info: RawTableInfo = new_test_table_info().into();
+        let region_route = new_test_region_route();
+        let region_routes = vec![region_route.clone()];
+        // creates metadata.
+        table_metadata_manager
+            .create_table_metadata(table_info.clone(), region_routes.clone())
+            .await
+            .unwrap();
+        let new_table_name = "another_name".to_string();
+        let table_info_value = TableInfoValue::new(table_info.clone());
+        table_metadata_manager
+            .rename_table(table_info_value.clone(), new_table_name.clone())
+            .await
+            .unwrap();
+        // if remote metadata was updated, it should be ok.
+        table_metadata_manager
+            .rename_table(table_info_value.clone(), new_table_name.clone())
+            .await
+            .unwrap();
+        let mut modified_table_info = table_info;
+        modified_table_info.name = "hi".to_string();
+        let modified_table_info_value = table_info_value.update(modified_table_info);
+        // if the table_info_value is wrong, it should return an error.
+        // The ABA problem.
```

Comment:
Why it's an "ABA" problem? The key seems didn't change to another value and back, during the txn.

Reply:
Yes, but the comparing value changed.